### PR TITLE
solution

### PR DESCRIPTION
**Pull request by abcdesign1**

### CODE REVIEW

File: src/averageAges.js [hunk 1]
```diff
@@ -14,13 +51,17 @@
...
+function calculateMenAverageAge(people, century = 21) {
+  const men = getMen(people);
+  const menDiedInCentury = men
+    .filter(man => getCentury(man['died']) === century);
+
+  const menAges = getAges(menDiedInCentury).length !== 0
+    ? getAges(menDiedInCentury) : getAges(men);
+
+  return getAverage(menAges);
+};
```
Comment (sergey-lipikhin) [inline src/averageAges.js:64]:
```suggestion
  const men = century === undefined
    ? getMen(people)
    : getMen(people)
      .filter(man => (
        getCentury(man['died']) === century
      ));

  const menAges = getAges(men);

  return getAverage(menAges);
```

Check if `century === undefined` on the first step of your algorithm

File: src/averageAges.js [hunk 2]
```diff
@@ -1,5 +1,42 @@
...
+    ? resultArr.push(person[key]) : null);
+
+  return resultArr;
+};
+
+const getMotherNames = getValuesArr('mother');
+
+const getMothers = (people, group = people) => {
+  const women = getWomen(people);
+  const motherNames = getMotherNames(group);
```
Comment (sergey-lipikhin) [inline src/averageAges.js:25]: You pass `group` into `getMotherNames`. But `getMotherNames` does not take parameters :)

File: src/averageAges.js [hunk 3]
```diff
@@ -1,5 +1,42 @@
...
+};
+
+const getCentury = died => Math.ceil(died / 100);
+
+const getAges = people => people.map(person => person['died'] - person['born']);
+
+const getAverage = ages => +(
+  ((ages.reduce((acc, cur) => acc + cur, 0)) / ages.length
+  ).toFixed(2)
+) || 0;
```
Comment (sergey-lipikhin) [inline src/averageAges.js:38]:
```suggestion
const getAverage = ages => (
  ages.reduce((acc, cur) => acc + cur, 0) / ages.length
);
```

File: src/averageAges.js [hunk 4]
```diff
@@ -36,8 +77,11 @@ function calculateMenAverageAge(people, century) {
  *
  * @return {number}
  */
-function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+function calculateWomenAverageAge(people, withChildren = false) {
+  const women = !withChildren ? getWomen(people) : getMothers(people);
+  const womenAges = getAges(women);
+
+  return getAverage(womenAges);
```
Comment (sergey-lipikhin) [inline src/averageAges.js:84]: 👍

File: src/averageAges.js [hunk 5]
```diff
@@ -1,5 +1,42 @@
...
+
+const getMotherNames = getValuesArr('mother');
+
+const getMothers = (people, group = people) => {
+  const women = getWomen(people);
+  const motherNames = getMotherNames(group);
+  const mothers = women.filter(woman => motherNames.includes(woman['name']));
+
+  return mothers;
+};
```
Comment (sergey-lipikhin) [inline src/averageAges.js:29]: All of this to select all mothers within people? I bet you can replace this with far less code lines.

1. Pick up all women like you did `const women = getWomen(people);` ✔️
2. Filter all women who has **AT LEAST** one child (think what the function to use)❌

```suggestion
const getMothers = (people, group = people) => {
  const women = getWomen(people);
  // filter women and get mothers
  
  return mothers;
};
```

**Review by sergey-lipikhin — CHANGES_REQUESTED** [review]

File: src/averageAges.js [hunk 6]
```diff
@@ -1,5 +1,42 @@
 'use strict';
 
+const getByKeyValue = (key, value) => people => {
+  const peopleCopy = [ ...people ];
```
[the remark this answers is not in the document]
Comment (vadym4che) [reply src/averageAges.js:4]: fixed (adding this statement was caused by panic with failing tests)))

File: src/averageAges.js [hunk 7]
```diff
@@ -14,13 +51,17 @@
...
  */
-function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
-}
+
+function calculateMenAverageAge(people, century = 21) {
```
[the remark this answers is not in the document]
Comment (vadym4che) [reply src/averageAges.js:55]: fixed (I've read task condition - as "If century is NOT defined it should be current" - I was inattentive in this point)
thank you

File: src/averageAges.js (hunk 1 above)
Comment (vadym4che) [reply src/averageAges.js:64]: done

File: src/averageAges.js (hunk 2 above)
Comment (vadym4che) [reply src/averageAges.js:25]: image
please be more attentive too

I have callback constructor `getValuesArr = (key) =>` that returns function `people => {}`, wose accept some array (called people) and reterns array of people

`getMotherNames(group)` is example of function, that was created by fuction constructor
and designed to accept an array of objects of people and return an array of string-kind names of mother

so consider written above and
thank you for your comments

File: src/averageAges.js (hunk 3 above)
Comment (vadym4che) [reply src/averageAges.js:38]: I reduced this code-line accordingly your suggestions
I had some problems with tests
and written this after I've seen test log like `expected: 67.68  ... received...`

I accepted your variant - it pass all test successfully

so thank you

File: src/averageAges.js (hunk 5 above)
Comment (vadym4che) [reply src/averageAges.js:29]: shortened a bit
thanks

File: src/averageAges.js [hunk 8]
```diff
@@ -54,8 +79,18 @@ function calculateWomenAverageAge(people, withChildren) {
...
-function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+
+function calculateAverageAgeDiff(people, onlyWithSon = false) {
+  const children = onlyWithSon ? getMen(people) : people;
+  const mothers = getMothers(people);
+  const ageDiffs = children.map(({ mother: motherName, born: childBorn }) => {
+    const mother = mothers.filter(({ name }) => name === motherName);
+
+    return mother.length !== 0 ? childBorn - mother[0]['born'] : null;
```
Comment (sergey-lipikhin) [inline src/averageAges.js:89]:
```suggestion
    return mother.length !== 0
      ? childBorn - mother[0]['born']
      : null;
```

File: src/averageAges.js [hunk 9]
```diff
@@ -36,8 +58,11 @@ function calculateMenAverageAge(people, century) {
  *
  * @return {number}
  */
-function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+function calculateWomenAverageAge(people, withChildren = false) {
+  const women = !withChildren ? getWomen(people) : getMothers(people);
```
Comment (sergey-lipikhin) [inline src/averageAges.js:62]:
```suggestion
  const women = !withChildren
    ? getWomen(people)
    : getMothers(people);
```

File: src/averageAges.js [hunk 10]
```diff
@@ -14,13 +33,16 @@
...
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
-}
+  const men = getMen(people);
+  const menDiedInCentury = men
+    .filter(man => getCentury(man['died']) === century);
+  const menAges = century
+    ? getAges(menDiedInCentury) : getAges(men);
```
Comment (sergey-lipikhin) [inline src/averageAges.js:42]:
```suggestion
  const menAges = century
    ? getAges(menDiedInCentury)
    : getAges(men);
```

File: src/averageAges.js [hunk 11]
```diff
@@ -1,5 +1,24 @@
...
+  return people.filter(person => person[key] === value);
+};
+
+const getMen = getByKeyValue('sex', 'm');
+const getWomen = getByKeyValue('sex', 'f');
+
+const getMothers = (people) => {
+  const women = getWomen(people);
+
+  return women.filter(woman => people.some(person => person.mother === woman.name));
```
Comment (sergey-lipikhin) [inline src/averageAges.js:13]:
```suggestion
  return women.filter(woman => (
    people.some(person => person.mother === woman.name)
  ));
```

**Review by sergey-lipikhin — CHANGES_REQUESTED** [review]
Please, fix indentations

**Review by sergey-lipikhin — APPROVED** [review]

File: src/averageAges.js (hunk 10 above)
Comment (illiarerun) [reply src/averageAges.js:43]: agree, format your ternaries properly

File: src/averageAges.js (hunk 9 above)
Comment (illiarerun) [reply src/averageAges.js:63]: agree

File: src/averageAges.js (hunk 8 above)
Comment (illiarerun) [reply src/averageAges.js:90]: agree

**Review by illiarerun — CHANGES_REQUESTED** [review]

**Comment by vadym4che** [comment]
fixed code style issues

File: src/averageAges.js [hunk 12]
```diff
@@ -1,5 +1,25 @@
...
+  return people.filter(person => person[key] === value);
+};
+
+const getMen = getByKeyValue('sex', 'm');
+const getWomen = getByKeyValue('sex', 'f');
+
+const getMothers = (people) => {
+  const women = getWomen(people);
+
+  return women
```
Comment (sergii-nosachenko) [inline src/averageAges.js:13]:
```suggestion
  return people
```
Mothers are always women, so you can optimize logic.

File: src/averageAges.js [hunk 13]
```diff
@@ -54,8 +83,22 @@ function calculateWomenAverageAge(people, withChildren) {
...
-function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+
+function calculateAverageAgeDiff(people, onlyWithSon = false) {
+  const children = onlyWithSon
+    ? getMen(people)
+    : people;
+  const mothers = getMothers(people);
+  const ageDiffs = children.map(({ mother: motherName, born: childBorn }) => {
+    const mother = mothers.filter(({ name }) => name === motherName);
```
Comment (sergii-nosachenko) [inline src/averageAges.js:93]:
```suggestion
    const mother = mothers.find(({ name }) => name === motherName);
```
If you need only first occurrence - use `find`

File: src/averageAges.js [hunk 14]
```diff
@@ -54,8 +83,22 @@ function calculateWomenAverageAge(people, withChildren) {
...
+
+    return mother.length !== 0
+      ? childBorn - mother[0]['born']
+      : null;
```
Comment (sergii-nosachenko) [inline src/averageAges.js:97]: Can be refactored if you use `find`

**Review by sergii-nosachenko — APPROVED** [review]
Good work 👍
Read comments.

**Pull request closed without merging**